src/rt/ds/hashmap.h

```diff
@@ -256,411 +195,272 @@ namespace verona::rt
         return entry().second;
       }
 
+      bool is_marked()
+      {
+        return key_of(entry()) & MARK_MASK;
+      }
+
+      void mark()
+      {
+        key_of(entry()) |= MARK_MASK;
+      }
+
       EntryView operator*()
       {
-        auto* key = (Object*)unmark_key(key_of(entry()));
         if constexpr (is_set)
-          return key;
+          return key();
         else
-          return std::make_pair(key, &entry().second);
+          return std::make_pair(key(), &value());
       }
 
       Iterator& operator++()
       {
-        assert(map->count > 0);
-
-        auto size = map->get_size();
-        while (true)
+        while (++index < map.capacity())
         {
-          i++;
-          if (i == size)
-            break;
-          if (key_of(map->set[i]) != 0)
+          const auto key = key_of(map.slots[index]);
+          if (key != 0)
             break;
         }
-
         return *this;
       }
 
-      size_t get_index()
+      bool operator==(const Iterator& other) const
       {
-        return i;
+        return (index == other.index) && (&map == &other.map);
       }
 
-      bool operator!=(const Iterator& it) const
+      bool operator!=(const Iterator& other) const
       {
-        return i != it.i;
-      }
-
-      bool operator==(const Iterator& it) const
-      {
-        return i == it.i;
+        return !(*this == other);
       }
     };
 
-    // The key returned has no mark bits or dib.
-    static uintptr_t unmark_key(uintptr_t p)
+    /**
+     * Create an `ObjectMap` with an initial capacity for at least 8 entries.
+     */
+    ObjectMap(Alloc* alloc)
     {
-      assert(p != 0);
-      return p & POINTER_MASK;
+      static constexpr size_t init_alloc =
+        bits::next_pow2_const(sizeof(Entry) * 8);
+      capacity_shift = bits::ctz(init_alloc / sizeof(Entry));
```

Comment:
As discussed privately.  This doesn't handle cases where `sizeof(Entry)` is not a power of two, and in the cases where it is, is overly complex.

I am happy for a `static_assert` to force power of 2 sizes. Or to add a rounding/alignment constraint to `Entry`.